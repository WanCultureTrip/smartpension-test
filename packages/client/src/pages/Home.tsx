import React, { useState } from 'react'
import type { FC, KeyboardEvent } from 'react'
import {
  Container,
  InputRightElement,
  Input,
  Heading,
  InputGroup,
  IconButton,
  VStack,
  Progress,
  Alert,
  AlertIcon,
} from '@chakra-ui/react'
import { Search2Icon } from '@chakra-ui/icons'
import { CityList } from '../components/CityList/CityList'
import { useCitySearch } from '../hooks/useCitySearch'

export const Home: FC = () => {
  const [searchTerm, setSearchTerm] = useState('')
  const { cities, loading, error, onSearch } = useCitySearch()

  const handleKeyDown = (event: KeyboardEvent) => {
    if (event.key === 'Enter') {
      onSearch(searchTerm)
    }
  }

  return (
    <VStack spacing="8">
      <Heading as="h1">Smart traveller</Heading>
      <Container maxW="container.md">
        <InputGroup>
          <Input
            aria-label="search-input"
            onKeyDown={handleKeyDown}
            onChange={e => setSearchTerm(e.target.value)}
            value={searchTerm}
          />
          <InputRightElement
            children={
              <IconButton onClick={() => onSearch(searchTerm)} aria-label="search-button" icon={<Search2Icon />} />
            }
          />
        </InputGroup>

        {loading && <Progress m="1rem" size="xs" isIndeterminate />}
        {error && (
          <Alert status="error">
            <AlertIcon />
            There was an error processing your request
          </Alert>
        )}
        <CityList cityList={cities} showEditMenu />
      </Container>
    </VStack>
  )
}
