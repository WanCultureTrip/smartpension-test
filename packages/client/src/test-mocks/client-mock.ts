import type { MockedResponse } from '@apollo/client/testing'
import { CITY_FILTER } from '../graphql/CITY_FILTER.query'
import { UPDATE_CITY } from '../graphql/UPDATE_CITY.query'
import type { CityType } from '../types/city-types'

const mockCityList = [
  {
    id: 1,
    name: 'London',
    visited: true,
    wishlist: true,
    country: 'United Kingdom',
  },
  {
    id: 14,
    name: 'Barcelona',
    visited: true,
    wishlist: false,
    country: 'Spain',
  },
  {
    id: 50,
    name: 'Lodz',
    visited: false,
    wishlist: false,
    country: 'Poland',
  },
  {
    id: 82,
    name: 'Oslo',
    visited: false,
    wishlist: false,
    country: 'Norway',
  },
]

const getMockResults = (cityList: CityType[]) => {
  return {
    data: {
      cities: {
        total: cityList.length,
        cities: cityList,
      },
    },
  }
}

export const clientMock = [
  {
    request: {
      query: CITY_FILTER,
      variables: {
        filter: { name: 'Lo' },
      },
    },
    result: getMockResults(mockCityList),
  },
  {
    request: {
      query: UPDATE_CITY,
      variables: {
        input: {
          visited: true,
          wishlist: false,
          id: 82,
        },
      },
    },
    result: { data: { updateCity: { ...mockCityList[3], visited: true } } },
  },
  {
    request: {
      query: CITY_FILTER,
      variables: {
        filter: { visited: true },
      },
    },
    result: getMockResults(mockCityList.filter(item => item.visited)),
  },
  {
    request: {
      query: CITY_FILTER,
      variables: {
        filter: { wishlist: true },
      },
    },
    result: getMockResults(mockCityList.filter(item => item.wishlist)),
  },
] as MockedResponse[]
