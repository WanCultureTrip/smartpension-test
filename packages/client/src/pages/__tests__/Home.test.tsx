import { render, fireEvent, waitFor } from '@testing-library/react'
import '@testing-library/jest-dom'
import { Home } from '../Home'
import { MockedProvider } from '@apollo/client/testing'
import { clientMock } from '../../test-mocks/client-mock'

const renderComp = () => {
  const utils = render(
    <MockedProvider mocks={clientMock} addTypename={false}>
      <Home />
    </MockedProvider>
  )
  const input = utils.getByLabelText('search-input') as HTMLInputElement
  return {
    input,
    ...utils,
  }
}

describe('<Home /> component', () => {
  test('should match initial snapshot', async () => {
    const { container } = renderComp()
    expect(container).toMatchSnapshot()
  })

  test('should search for cities', async () => {
    const { input, getByLabelText, getAllByLabelText, findByText } = renderComp()

    // put Lo into the search input field
    fireEvent.change(input, { target: { value: 'Lo' } })
    expect(input.value).toBe('Lo')

    // click on the search button
    fireEvent.click(getByLabelText('search-button'))

    // wait for cards to load
    await waitFor(() => getAllByLabelText('city-card'))

    // check that 4 city cards is loaded
    expect(getAllByLabelText('city-card').length).toBe(4)

    // check that one of the cities is London
    expect(findByText('London')).toBeDefined()
  })

  test('should search for cities and click on visited and wish list', async () => {
    const { input, getByLabelText, getAllByLabelText, findAllByLabelText } = renderComp()

    // put Lo into the search input field
    fireEvent.change(input, { target: { value: 'Lo' } })
    expect(input.value).toBe('Lo')

    // click on the search button
    fireEvent.click(getByLabelText('search-button'))

    // wait for cards to load
    await waitFor(() => getAllByLabelText('city-card'))

    // check that 4 city cards is loaded
    expect(getAllByLabelText('city-card').length).toBe(4)

    // check that one of the cities is London
    const allVisitedCheckbox = await findAllByLabelText('visited-checkbox')
    expect(allVisitedCheckbox).toBeDefined()

    // last city should have visited false
    const firstVisitedCheckbox = allVisitedCheckbox[3] as HTMLInputElement
    expect(firstVisitedCheckbox.checked).toBe(false)

    // click the first city checkbox
    fireEvent.click(firstVisitedCheckbox)
  })
})
