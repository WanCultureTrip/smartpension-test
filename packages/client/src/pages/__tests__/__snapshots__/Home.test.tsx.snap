// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> component should match initial snapshot 1`] = `
<div>
  <div
    class="chakra-stack css-14witgw"
  >
    <h1
      class="chakra-heading css-0"
    >
      Smart traveller
    </h1>
    <div
      class="chakra-container css-gygvqc"
    >
      <div
        class="chakra-input__group css-4302v8"
      >
        <input
          aria-label="search-input"
          class="chakra-input css-0"
          value=""
        />
        <div
          class="chakra-input__right-element css-dp9r5f"
        >
          <button
            aria-label="search-button"
            class="chakra-button css-198bfjk"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="chakra-icon css-onkibi"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M23.414,20.591l-4.645-4.645a10.256,10.256,0,1,0-2.828,2.829l4.645,4.644a2.025,2.025,0,0,0,2.828,0A2,2,0,0,0,23.414,20.591ZM10.25,3.005A7.25,7.25,0,1,1,3,10.255,7.258,7.258,0,0,1,10.25,3.005Z"
                fill="currentColor"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        aria-label="city-list"
        class="css-63tcgz"
      />
    </div>
  </div>
</div>
`;
